docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
+| Read or set variables | Y | Y | 
+| set restricted variables(some of them even can not be read) | N | Y |
+| Read or set restricted system tables | N | Y | 
+| DROP USER cloudAdmin | N | Hard to N(unless some hardcoded) |
+| REVOKE cloudAdmin | N | Hard to N(unless some hardcoded) |
+| Show processlist, Access to threads belong to other users | Y | Y |
+| Change password if the password of SUPER user is forgotten | N | Y |
+| Kill connections belong to cloudAdmin | N | Y |
+| SHUTDOWN / RESTART | N | Y (graceful shutdown on k8s for tidb-server) |
+
+### Compatibility Tests
+
+The introduction of `DYNAMIC` privileges is not expected to introduce any compatibility issues, because backwards compatibility is ensured. However, plugins should migrate to registering their own dynamic privileges and not rely on the use of `SUPER`. This is considered an enhancement, and not included in-scope for the initial introduction of dynamic privileges (which introduces the framework for plugins to use).
+
+## Impact & Risks
+
+In its initial release, dynamic privilege usage will be controlled by an experimental feature flag. The implementation will be via restricting `GRANT` and `REVOKE` statements from creating dynamic privileges (it is too invasive to conditionally modify the ast visitor functionaliy).
+
+For backwards compatibility, the MySQL-compatible dynamic privileges will also permit `SUPER`. This helps prevent upgrade issues, such a when TiDB was bootstrapped `GRANT ALL ON *.*` would not have granted all the dynamic privileges. There might be some impact on Upgrade/Downgrade story if eventually the `BACKUP_ADMIN` privilege is used instead of `SUPER`, but for the initial release I am planning to allow either.
+
+## Investigation & Alternatives
+
+An alternative could be to support fine-grained access in a TiDB specific way. Because the MySQL functionality overlaps nicely, it doesn’t really make sense not to follow.
```

Comment:
can we solve the problem by role-based access control? I mean, can we define several roles have different privileges to solve the problems in DBaaS?

Reply:
RBAC is orthoginal. This is about the fine-grained level of what can be assigned. RBAC is about how to manage those credentials.

Dynamic privileges works with RBAC. Here is a testcase which shows the relationship where dynamic privileges are assigned to a role, and then the role is granted:

https://github.com/pingcap/tidb/pull/22778/files#diff-8b507d8e9d7f506b565cbedfd8e7c860aa76099537d36ed9cdeaf931d71d64c0R1200-R1243

Reply:
It can't be solved purely via RBAC because the MySQL rules are that you can only assign a privilege for which you have yourself (+ you have the grant option, which in static privileges is only global). So static privileges would never allow a case where you are a role admin that can assign new permissions to someone, but you can't view data yourself. 

In Dynamic privileges, the `GRANT OPTION` is moved to each privilege. Or to paraphrase: it can't be solved by RBAC because static privileges are far too limited.